Inconsistent behaviour of test_extra_env_vars with docker environments

**Describe the bug**
I have configured the following env vars:

In the pants.toml
```toml
[test]
extra_env_vars = [
  "ENV_TOML=toml",
  "ENV_ALL=toml"
]
```

In the docker_environment
```python
docker_environment(
    name="dev_docker",
    platform="linux_x86_64",
    image="python:3.10.4-slim-bullseye",
    python_bootstrap_search_path=["<PATH>"],
    test_extra_env_vars=[
        "ENV_DOCKER=docker",
        "ENV_ALL=docker"
    ],
)
```

In the test target
```python
python_tests(
    environment="dev",
    extra_env_vars=[
        "ENV_TARGET=target",
        "ENV_ALL=target"
    ],
 )
```

My assumption would be that the following test passes when executed in the docker_environment:
```python
import os

def test_env_var():
    envs = {k: v for k,v in os.environ.items() if k.startswith('ENV_')}
    assert envs["ENV_TARGET"] == 'target'
    assert envs["ENV_ALL"] == 'target'

    assert envs["ENV_TOML"] == 'toml'
    assert envs["ENV_DOCKER"] == 'docker'
```

But I get the following errors:

```shell
➜ pants test tests/python/hello_world/messages/test_env_vars.py         
11:52:25.24 [ERROR] Completed: Run Pytest - (environment:dev, tests/python/hello_world/messages/test_env_vars.py) - failed (exit code 1).
============================= test session starts ==============================
platform linux -- Python 3.10.4, pytest-7.0.1, pluggy-1.0.0
rootdir: /pants-sandbox/pants-sandbox-uh2dCY
plugins: cov-3.0.0, forked-1.4.0, xdist-2.5.0
collected 1 item

tests/python/hello_world/messages/test_env_vars.py F                     [100%]

=================================== FAILURES ===================================
_________________________________ test_env_var _________________________________

    def test_env_var():
        envs = {k: v for k,v in os.environ.items() if k.startswith('ENV_')}
        assert envs["ENV_TARGET"] == 'target'
        assert envs["ENV_ALL"] == 'target'
    
>       assert envs["ENV_TOML"] == 'toml'
E       KeyError: 'ENV_TOML'
```

When I run the same test with the `--debug` flag I get a different error:

```shell
➜ pants test --debug tests/python/hello_world/messages/test_env_vars.py 
========================================== test session starts ==========================================
platform darwin -- Python 3.10.9, pytest-7.0.1, pluggy-1.0.0
rootdir: /private/var/folders/8b/0w6lmjjx7fsdm_6hhcrjfdfw0000gq/T/pants-sandbox-0SBEZ4
plugins: xdist-2.5.0, forked-1.4.0, cov-3.0.0
collected 1 item                                                                                        

tests/python/hello_world/messages/test_env_vars.py F                                              [100%]

=============================================== FAILURES ================================================
_____________________________________________ test_env_var ______________________________________________

    def test_env_var():
        envs = {k: v for k,v in os.environ.items() if k.startswith('ENV_')}
        assert envs["ENV_TARGET"] == 'target'
        assert envs["ENV_ALL"] == 'target'
    
        assert envs["ENV_TOML"] == 'toml'
>       assert envs["ENV_DOCKER"] == 'docker'
E       KeyError: 'ENV_DOCKER'
```

How is pants supposed to handle the env vars that are configured on the different levels? Are the lists supposed to be merged together or do should they be completely be replaced by the deepest configuration?

**Pants version**
2.16.0.dev5, we are using the new scie-pants binary in version 0.3.2

**OS**
MacOS

**Additional info**
I have created a reproducible example here https://github.com/psontag/pexample/tree/pants-env-vars

--- a/src/python/pants/core/goals/test.py
+++ b/src/python/pants/core/goals/test.py
@@ -9,7 +9,7 @@ from abc import ABC, ABCMeta
 from dataclasses import dataclass
 from enum import Enum
 from pathlib import PurePath
-from typing import Any, ClassVar, Iterable, Optional, TypeVar, cast
+from typing import Any, ClassVar, Iterable, Optional, Sequence, TypeVar, cast
 
 from pants.core.goals.multi_tool_goal_helper import SkippableSubsystem
 from pants.core.goals.package import BuiltPackage, EnvironmentAwarePackageRequest, PackageFieldSet
@@ -732,28 +732,26 @@ async def _get_test_batches(
 
 async def _run_debug_tests(
     batches: Iterable[TestRequest.Batch],
+    environment_names: Sequence[EnvironmentName],
     test_subsystem: TestSubsystem,
     debug_adapter: DebugAdapterSubsystem,
-    local_environment_name: ChosenLocalEnvironmentName,
 ) -> Test:
-    # TODO: Because these are interactive, they are always pinned to the local environment.
-    # See https://github.com/pantsbuild/pants/issues/17182
     debug_requests = await MultiGet(
         (
             Get(
                 TestDebugRequest,
-                {batch: TestRequest.Batch, local_environment_name.val: EnvironmentName},
+                {batch: TestRequest.Batch, environment_name: EnvironmentName},
             )
             if not test_subsystem.debug_adapter
             else Get(
                 TestDebugAdapterRequest,
-                {batch: TestRequest.Batch, local_environment_name.val: EnvironmentName},
+                {batch: TestRequest.Batch, environment_name: EnvironmentName},
             )
         )
-        for batch in batches
+        for batch, environment_name in zip(batches, environment_names)
     )
     exit_code = 0
-    for debug_request in debug_requests:
+    for debug_request, environment_name in zip(debug_requests, environment_names):
         if test_subsystem.debug_adapter:
             logger.info(
                 softwrap(
@@ -768,7 +766,7 @@ async def _run_debug_tests(
             InteractiveProcessResult,
             {
                 debug_request.process: InteractiveProcess,
-                local_environment_name.val: EnvironmentName,
+                environment_name: EnvironmentName,
             },
         )
         if debug_result.exit_code != 0:
@@ -817,11 +815,6 @@ async def run_tests(
         test_subsystem,
     )
 
-    if test_subsystem.debug or test_subsystem.debug_adapter:
-        return await _run_debug_tests(
-            test_batches, test_subsystem, debug_adapter, local_environment_name
-        )
-
     environment_names = await MultiGet(
         Get(
             EnvironmentName,
@@ -830,6 +823,11 @@ async def run_tests(
         )
         for batch in test_batches
     )
+
+    if test_subsystem.debug or test_subsystem.debug_adapter:
+        return await _run_debug_tests(
+            test_batches, environment_names, test_subsystem, debug_adapter
+        )
 
     results = await MultiGet(
         Get(TestResult, {batch: TestRequest.Batch, environment_name: EnvironmentName})

--- a/src/python/pants/core/goals/test_integration_test.py
+++ b/src/python/pants/core/goals/test_integration_test.py
@@ -1,0 +1,47 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from textwrap import dedent
+
+from pants.testutil.pants_integration_test import PantsResult, run_pants, setup_tmpdir
+
+
+def test_environment_usage() -> None:
+    files = {
+        "project/tests.py": dedent(
+            """\
+            def test_thing():
+                pass
+            """
+        ),
+        "project/BUILD": "python_tests(environment='python_38')",
+        "BUILD": dedent(
+            """\
+            docker_environment(
+                name="python_38",
+                image="python:3.8",
+                python_bootstrap_search_path=["<PATH>"],
+            )
+            """
+        ),
+    }
+
+    with setup_tmpdir(files) as dirname:
+
+        def run(*extra_test_args: str) -> PantsResult:
+            return run_pants(
+                [
+                    "--backend-packages=['pants.backend.python']",
+                    "--python-interpreter-constraints=['==3.8.*']",
+                    f"--environments-preview-names={{'python_38': '{dirname}:python_38'}}",
+                    "test",
+                    *extra_test_args,
+                    f"{dirname}/project:",
+                ],
+            )
+
+        # A normal run should succeed.
+        run().assert_success()
+
+        # A debug run should fail (TODO: currently, see #17182).
+        run("--debug").assert_failure()

--- a/src/python/pants/core/goals/test_integration_test.py
+++ b/src/python/pants/core/goals/test_integration_test.py
@@ -44,4 +44,6 @@ def test_environment_usage() -> None:
         run().assert_success()
 
         # A debug run should fail (TODO: currently, see #17182).
-        run("--debug").assert_failure()
+        debug_run = run("--debug")
+        debug_run.assert_failure()
+        assert "Only local environments support running processes interactively" in debug_run.stderr

--- a/src/rust/engine/src/intrinsics.rs
+++ b/src/rust/engine/src/intrinsics.rs
@@ -535,7 +535,16 @@ fn interactive_process(
         (py_interactive_process.extract().unwrap(), py_process, process_config)
       });
       match process_config.environment.strategy {
-        ProcessExecutionStrategy::Docker(_) | ProcessExecutionStrategy::RemoteExecution(_) => Err("InteractiveProcess should not set docker_image or remote_execution".to_owned()),
+        ProcessExecutionStrategy::Docker(_) | ProcessExecutionStrategy::RemoteExecution(_) => {
+          // TODO: #17182 covers adding support for running processes interactively in Docker.
+          Err(
+            format!(
+              "Only local environments support running processes \
+               interactively, but a {} environment was used.",
+              process_config.environment.strategy.strategy_type(),
+            )
+          )
+        },
         _ => Ok(())
       }?;
       let mut process = ExecuteProcess::lift(&context.core.store(), py_process, process_config).await?.process;

